### ISSUE TITLE
avx implementation of convolution pad in jit

### ISSUE THREAD

**Issue by kangkangfs** [oneapi-src/oneDNN#1098]
Dear sirs,
In the init_conf function in jit_avx2_conv_kernel_f32.cpp at line 766 , when jcp.l_pad=3, the avx implementation will trigger an assertion error at line 796. How can i deal with it?

    if (jcp.ow < jcp.ur_w) jcp.ur_w = jcp.ow;
    jcp.ur_w_tail = jcp.ow % jcp.ur_w;

    args_ok = true && IMPLICATION(!is_data_layout_nxc, jcp.oc % simd_w == 0)
            && jcp.l_pad <= jcp.ur_w
            && IMPLICATION(jcp.kw > 7,
                    (jcp.t_pad == 0 && jcp.l_pad == 0)
                            || (jcp.stride_w == 1 && jcp.stride_h == 1))
            && IMPLICATION(mimo && !is_data_layout_nxc, jcp.ic % simd_w == 0);
    if (!args_ok) return status::unimplemented;

    jcp.ic_tail = is_data_layout_nxc ? jcp.ic % simd_w : 0;
    jcp.oc_tail = is_data_layout_nxc
            ? jcp.oc % simd_w
            : (jcp.with_binary ? jcp.oc_without_padding % simd_w : 0);

    int r_pad_no_tail = nstl::max(0,
            calculate_end_padding(jcp.l_pad, jcp.ow - jcp.ur_w_tail, jcp.iw,
                    jcp.stride_w, ext_kw));

    if (r_pad_no_tail > jcp.ur_w * jcp.stride_w && jcp.ow / jcp.ur_w > 1) {
        /* recalculate ur_w, nb_oc_blocking and ur_w_tail */
        jcp.ur_w = nstl::min(r_pad_no_tail / jcp.stride_w + jcp.ur_w_tail,
                nstl::min(jcp.ow, num_avail_regs / 2));
        jcp.nb_oc_blocking = (num_avail_regs - jcp.ur_w) / jcp.ur_w;
        jcp.ur_w_tail = jcp.ow % jcp.ur_w;
        /* check again ... */
        r_pad_no_tail = nstl::max(0,
                calculate_end_padding(jcp.l_pad, jcp.ow - jcp.ur_w_tail, jcp.iw,
                        jcp.stride_w, ext_kw));
        if (jcp.ur_w < nstl::max(jcp.l_pad, r_pad_no_tail))
            return status::unimplemented;
    }
    assert(jcp.nb_oc_blocking > 0);
    assert(jcp.ur_w * (jcp.nb_oc_blocking + 1) <= num_avail_regs);

**Comment by kwiersch** [comment]
Hi @kangkangfs, could you please share the exact convolution shape that triggers the assert? I ran a simple test case using [benchdnn](https://github.com/oneapi-src/oneDNN/blob/master/tests/benchdnn/README.md) but did not see any asserts:
```bash
$ ./benchdnn -v5 --conv ic64oc64ih56kh7ph3
run: --conv ic64ih56oc64oh56kh7ph3
oneDNN implementation: jit:avx
oneDNN implementation: jit:avx
@@@ [DST] final: diff: err:0, l0(``-0``) l1:(2.59927e+09,2.59927e+09,0,``0``) l2:(5.01541e+06,5.01541e+06,0,``0``) l8:(26272,26272,-0,``-0``)
0:PASSED __REPRO: --conv ic64ih56oc64oh56kh7ph3
tests:1 passed:1 skipped:0 mistrusted:0 unimplemented:0 failed:0 listed:0
```

**Comment by kangkangfs** [comment]
Thanks for your reply!  for example, when pad=3, output_channel<=8, kernel=3, stride=1, the assertion error will be triggered.

**Comment by kwiersch** [comment]
When the kernel size is 3, padding of 3 or larger will not dispatch to jit implementation -- see [src/cpu/x64/jit_avx2_conv_kernel_f32.cpp#L642](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/jit_avx2_conv_kernel_f32.cpp#L642). So I don't think the assertion could be triggered in that case.

By the way, if you have observed the assert while running oneDNN, you could try to use [DNNL_VERBOSE=2](https://oneapi-src.github.io/oneDNN/dev_guide_verbose.html) and share the verbose logs (there should be an entry for the last created convolution primitive, which we can then use to reproduce the issue).

**Comment by kangkangfs** [comment]
The condition that will trigger an assertion error is: pad=3 && output_channel<=8 && kernel_size=3 && stride=1. Adding MKLDNN_VERBOSE=1 or DNNL_VERBOSE=2 did not display its parameter information ,it seems that before these two instructions were recognized the assertion error has been trigged
![image](https://user-images.githubusercontent.com/86177991/123956038-db49a000-d9dc-11eb-8570-723be49ff98b.png)

**Comment by kwiersch** [comment]
Hi @kangkangfs, thank you for including a screen shot!

From the file path, it looks like you are using an older version of oneDNN (back when it was called MKL-DNN). Is it possible for you to update to latest version? Be warned that there are changes to API and ABI since MKL-DNN days, but also lots of bug fixes and performance improvements.

If you cannot update, then could you share the commit hash for the version you are using? That will allow me to investigate and find a solution for the bug you found.

On the other hand, if you are able to change source of MKL-DNN and rebuild, I think adding [a check such as this](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/jit_avx2_conv_kernel_f32.cpp#L639-L642) will fix the issue. Initially the jit kernels for convolution assumed that padding would always be less than the filter size, but did not always check that this was the case.

**Comment by kangkangfs** [comment]
i may not update the onednn at present, and I can not find the commit hash you want, the version I am using is mkldnn 0.21. According to your prompt, I added a conditional judgment before the assertion, which temporarily avoided this error. Maybe the conditional judgment I added was incomplete, which caused the conv that was able to go to the jit branch normally, now it goes to the gemm branch, the running speed is slower. I hope the version I provided can help you troubleshoot the problem, thanks.
